src/main/cpp/win.cpp

```diff
@@ -1133,4 +1133,111 @@ Java_net_rubygrapefruit_platform_internal_jni_WindowsRegistryFunctions_getValueN
     return true;
 }
 
+// File watching
+
+JavaVM* jvm = NULL;
+
+typedef struct watch_details {
+    HANDLE watchHandle;
+    HANDLE threadHandle;
+    jstring path;
+    jobject watcherCallback;
+} watch_details_t;
+
+DWORD WINAPI EventProcessingThread(LPVOID data) {
+    watch_details_t *details = (watch_details_t*) data;
+    while (TRUE) {
+        if (WaitForSingleObject(details->watchHandle, INFINITE) == WAIT_FAILED) {
+            // TODO Error handling
+            // mark_failed_with_errno(env, "could not wait for change notification", result);
+            break;
+        }
+        if (!FindNextChangeNotification(details->watchHandle)) {
+            if (GetLastError() == ERROR_INVALID_HANDLE) {
+                // Assumed closed
+                break;
+            }
+            // TODO Error handling
+            // mark_failed_with_errno(env, "could not schedule next change notification", result);
+            break;
+        }
+
+        JNIEnv* env;
+        int getEnvStat = jvm->GetEnv((void **)&env, JNI_VERSION_1_6);
+        if (getEnvStat == JNI_EDETACHED) {
+            int attachThreadStat = jvm->AttachCurrentThread((void **) &env, NULL);
+            if (attachThreadStat != JNI_OK) {
+                printf("~~~~ Problem with AttachCurrentThread: %d\n", attachThreadStat);
+                // TODO Error handling
+                // invalidStateDetected = true;
+                break;
+            }
+        } else if (getEnvStat == JNI_EVERSION) {
+            printf("~~~~ Problem with GetEnv: %d\n", getEnvStat);
+            // TODO Error handling
+            // invalidStateDetected = true;
+            break;
+        }
+
+        wchar_t* pathStr = java_to_wchar_path(env, details->path, NULL);
+        printf("~~~~ Changes: %ls\n", pathStr);
+        free(pathStr);
+
+        jclass callback_class = env->GetObjectClass(details->watcherCallback);
+        jmethodID methodCallback = env->GetMethodID(callback_class, "pathChanged", "(Ljava/lang/String;)V");
+        env->CallVoidMethod(details->watcherCallback, methodCallback, details->path);
+    }
+    return 0;
+}
+
+JNIEXPORT jobject JNICALL
+Java_net_rubygrapefruit_platform_internal_jni_WindowsFileEventFunctions_startWatch(JNIEnv *env, jclass target, jstring path, jobject javaCallback, jobject result) {
+    wchar_t* pathStr = java_to_wchar_path(env, path, result);
+    HANDLE watchHandle = FindFirstChangeNotificationW(
```

Comment:
From https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findfirstchangenotificationw:
"In the ANSI version of this function, the name is limited to MAX_PATH characters. To extend this limit to 32,767 wide characters, call the Unicode version of the function and prepend "\?" to the path. For more information, see Naming a File."
Is the current version limited to 256 chars?

Reply:
We are dropping this API, but yes, we need to add tests with long file paths.